Update parse_options to use argparse instead of the deprecated optparse.

stevenj opened the pull request:
Fix 'extra_options' so that meaningful and complete command line options
can be added on invocation.

fixes #54 

extra command arguments is strange in existing fabricate.py and i couldnt find a useful example to use them.  Also, optparse is deprecated.  So in making the extra_options actually useful, i also updates parse_options to argparse.

To use extra_options one needs to pass an array of dictionaries as the parameter to extra_options, like this:

```
EXTRA_OPTIONS = [
    {'name_or_flags':['-b','--bdir'],'action':'store','default':None,'help':'Override the default Build Directory'},
]

main(extra_options=EXTRA_OPTIONS)
```

the reason for this is that argparse.add_argument (and optparse.make_option before it) require keyworded parameters to be able to fully use them.  The keywords are expressed as the keys to the dictionary, and are automatically extracted into an option list using *\* argument expansion.

**'name_or_flags'** is special, it is for un-keyworded arguments, to define the name or flag of the option, its value is an array of strings, specifying the **'name'** or the **'flags'** or both.  Again it is turned into an argument using \* argument expansion.

Im not 100% happy with this method, as its a bit obtuse, but i couldn't see a more straightforward way to achieve it.  I don't think this is backward compatible with the old use of 'extra_options' but as far as i could tell it didn't have any good use as it was.

to use your new option, just refer to `main.options.\<option\>` in the example above that would be:
`main.options.bdir`

stevenj committed:
Update parse_options to use argparse instead of the deprecated optparse.
Fix 'extra_options' so that meaningful and complete command line options
can be added on invocation.

stevenj committed:
Update fabricate.py

Remove extraneous test print statement

pjz commented:
:+1:

stevenj committed:
Add __init__ so we can use directly as a module

rmatev commented:
Can we have this merged?

By the way, I don't see how this solves #54. I suspect the OP there wants something along the lines of this
```
python build.py run arg1 --release --named=value
```
being equivalent to this
```
python build.py 'run("arg1", release=True, named="value")'
```

chriscz commented:
`argparse` is not available in Python 2.6 (which fabricate supports).  We could introduce a dependency on this [argparse](pypi.python.org/pypi/argparse) package, but then fabricate has dependencies outside the standard library.

Instead I propose we fix the existing bugs exposed by the test suite, merge this update, and release fabricate 2.0. What do you think @rmatev?

rmatev commented:
@chriscz agreed. Do we drop 2.6 support after fabricate 2.0?

chriscz commented:
Yes, this will be part of 2.0 at which point we will drop 2.6 support.

stevenj committed:
Upstream Merge

stevenj committed:
cleanup python3/2 mess only support python 3.4+
base runner class missing _builder.  add init for it and
propogate changes to subclasses to call base class init.
refix args after I broke it during upstream merge.

stevenj committed:
Add utility functions which make writing build scripts much cleaner

stevenj committed:
Add example build script for embedded systems.

stevenj committed:
preserve vscode development state

stevenj committed:
Merge branch 'master' of github.com:stevenj/fabricate

stevenj committed:
Add utility scripts

stevenj committed:
Various code changes

stevenj committed:
Add local controllable copy of uf2conv to make bootloader files

stevenj committed:
Correct linter problems
add ability to generate BIN files and UF2 files for bootloader

stevenj committed:
fix lint issue with docstring

stevenj committed:
make fabricate track openat syscalls to get dependencies properly

stevenj committed:
Update README.md

chriscz commented:
Hey @stevenj It looks like you've added quite few unrelated scripts to this PR. Would you mind if I just merge up to the replacement of `optparse` with argparse?

stevenj commented:
no, i don't mind at all.
